### ISSUE TITLE
Does neon support tesla k80

### ISSUE THREAD

**Issue by fjanoos** [NervanaSystems/neon#200]
hi,
Does the current version of neon support Tesla K80 GPUs ?
Thanks.

**Comment by nervanasys** [comment]
Yes, Kepler support is back in neon as of v1.2, 85fd5c1. See #80 and http://www.nervanasys.com/neon-v1-2-release-kepler-aws-support-are-back-deep-resnets-and-more/

**Comment by fjanoos** [comment]
Great ! 
I'm trying out neon on a Linux 3.16 machine with 4x K80's and I'm getting the following error. Please advise.

```python
from neon.backends import gen_backend
be = gen_backend(backend='gpu', batch_size=128, rng_seed=1, stochastic_round=False)
```

```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-13-b4c3bcb788a4> in <module>()
      1 # setup the backedn
      2 from neon.backends import gen_backend
----> 3 be = gen_backend(backend='gpu', batch_size=128, rng_seed=1, stochastic_round=False)

/home/firdaus/.conda/envs/neon/lib/python2.7/site-packages/neon-1.2.1-py2.7.egg/neon/backends/__init__.pyc in gen_backend(backend, rng_seed, datatype, batch_size, stochastic_round, device_id, max_devices, compat_mode, deterministic_update)
     89         if gpuflag is False:
     90             raise RuntimeError("Device " + str(device_id) + " does not have CUDA compute " +
---> 91                                "capability 3.0 or greater")
     92         if backend == 'gpu':
     93             from neon.backends.nervanagpu import NervanaGPU

RuntimeError: Device 0 does not have CUDA compute capability 3.0 or greater


```

**Comment by nervanasys** [comment]
Can you add a print before line 88 for check_gpu.get_compute_capability(device_id) ?

Were the CUDA drivers installed at the time of neon installation, and does your PATH include `/usr/local/cuda/bin` and LD_LIBRARY_PATH include `/usr/local/cuda/lib` ?

**Comment by fjanoos** [comment]
No - I'm using a remote cloud machine and I don't think cuda was installed (at least the above path's don't exist). Is there some documentation on what pre-reqs are needed to run neon with gpus ?
Thanks !

**Comment by nervanasys** [comment]
You would need the [CUDA SDK and drivers](https://developer.nvidia.com/cuda-downloads). 
General docs on installation are [here](http://neon.nervanasys.com/docs/latest/user_guide.html#installation).

**Comment by fjanoos** [comment]
Hi,
I added cuda/lib64 to LD_LIBRARY_PATH  and cuda/bin to PATH (for cuda sdk version 7.0). Also, I added the print statement on line 88 of __init__.py.


this is the result:

```python
# setup the backedn
from neon.backends import gen_backend
be = gen_backend(backend='gpu', batch_size=128, rng_seed=1, stochastic_round=False)
```

```
0

---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-1-b4c3bcb788a4> in <module>()
      1 # setup the backedn
      2 from neon.backends import gen_backend
----> 3 be = gen_backend(backend='gpu', batch_size=128, rng_seed=1, stochastic_round=False)

/home/firdaus/.conda/envs/neon/lib/python2.7/site-packages/neon-1.2.1-py2.7.egg/neon/backends/__init__.py in gen_backend(backend, rng_seed, datatype, batch_size, stochastic_round, device_id, max_devices, compat_mode, deterministic_update)
     90         if gpuflag is False:
     91             raise RuntimeError("Device " + str(device_id) + " does not have CUDA compute " +
---> 92                                "capability 3.0 or greater")
     93         if backend == 'gpu':
     94             from neon.backends.nervanagpu import NervanaGPU

RuntimeError: Device 0 does not have CUDA compute capability 3.0 or greater

```
Please advise.

**Comment by ursk** [comment]
Can you run the following checks to make sure the GPUs, Cuda and PyCuda are being recognized correctly?
`nvidia-smi`
`nvcc --version`
`python -m pycuda.autoinit`
For the last one make sure that the neon virtualenv has been activated.

**Comment by fjanoos** [comment]
```
>nvidia-smi                                                        [0/9941]
Thu Feb 11 01:02:48 2016
+------------------------------------------------------+
| NVIDIA-SMI 346.89     Driver Version: 346.89         |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla K80           Off  | 0000:05:00.0     Off |                    0 |
| N/A   38C    P0    61W / 149W |     55MiB / 11519MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   1  Tesla K80           Off  | 0000:06:00.0     Off |                    0 |
| N/A   36C    P0    73W / 149W |     55MiB / 11519MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   2  Tesla K80           Off  | 0000:84:00.0     Off |                    0 |
| N/A   40C    P0    57W / 149W |     55MiB / 11519MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   3  Tesla K80           Off  | 0000:85:00.0     Off |                    0 |
| N/A   35C    P0    72W / 149W |     55MiB / 11519MiB |     96%      Default |
+-------------------------------+----------------------+----------------------+
```

```
>nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2015 NVIDIA Corporation
Built on Mon_Feb_16_22:59:02_CST_2015
Cuda compilation tools, release 7.0, V7.0.27
```

```
 >python -m pycuda.autoinit
/home/firdaus/.conda/envs/neon/bin/python: No module named pycuda
```

**Comment by ursk** [comment]
Something went wrong during installation, and the pycuda package, which is a dependency for neon, did not get installed. It looks like you are using anaconda python, which does not support the standard virtualenv installation. There is a community contributed neon install script for anaconda you can find through the install docs at http://neon.nervanasys.com/docs/latest/user_guide.html#anaconda 

Hope that helps!

**Comment by magnord** [comment]
I have the same problem (but with a GTX980). I'm using Anaconda on Ubuntu 14.04. I've followed the Anaconda installation instructions, but they seem to be for Mac OS X, as the dependencies contain OS X-only packages (like appnope). Also, some of the dependencies in neon-conda-environment.yaml are referring to specific package versions that are no longer available. 

After some changes I got the installation running:

```
(neon)neon>python setup.py develop
running develop
running egg_info
creating neon.egg-info
writing neon.egg-info/PKG-INFO
writing top-level names to neon.egg-info/top_level.txt
writing dependency_links to neon.egg-info/dependency_links.txt
writing manifest file 'neon.egg-info/SOURCES.txt'
reading manifest file 'neon.egg-info/SOURCES.txt'
writing manifest file 'neon.egg-info/SOURCES.txt'
running build_ext
Creating /home/manordin/anaconda2/envs/neon/lib/python2.7/site-packages/neon.egg-link (link to .)
Adding neon 1.2.1 to easy-install.pth file
Installing neon script to /home/manordin/anaconda2/envs/neon/bin
Installing nvis script to /home/manordin/anaconda2/envs/neon/bin

Installed /home/manordin/neon
Processing dependencies for neon==1.2.1
Finished processing dependencies for neon==1.2.1
```

I can run examples using CPU but not GPU.

```
examples>nvidia-smi 
Thu Feb 11 11:54:18 2016       
+------------------------------------------------------+                       
| NVIDIA-SMI 352.79     Driver Version: 352.79         |                       
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GTX 980     Off  | 0000:02:00.0      On |                  N/A |
| 32%   39C    P8    16W / 180W |    381MiB /  4086MiB |      1%      Default |
+-------------------------------+----------------------+----------------------+
```

```
(neon)examples>nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2015 NVIDIA Corporation
Built on Tue_Aug_11_14:27:32_CDT_2015
Cuda compilation tools, release 7.5, V7.5.17
```

```
(neon)examples>python -m pycuda.autoinit
/home/manordin/anaconda2/envs/neon/bin/python: No module named pycuda
```

**Comment by fjanoos** [comment]
@ursk I followed those instructions initially. However, during installation A lot of packages didn't get resolved (eg appnode, astroid). Also, yesterday I tried to explicitly install pycuda, but because I'm behind a firewall, pycuda has issues with the proxy server. Not sure how to proceed.

```

creating dist
creating 'dist/pycuda-2015.1.3-py2.7-linux-x86_64.egg' and adding 'build/bdist.linux-x86_64/egg' to it
removing 'build/bdist.linux-x86_64/egg' (and everything under it)
Processing pycuda-2015.1.3-py2.7-linux-x86_64.egg
creating /srv/home/firdaus/.conda/envs/neon/lib/python2.7/site-packages/pycuda-2015.1.3-py2.7-linux-x86_64.egg
Extracting pycuda-2015.1.3-py2.7-linux-x86_64.egg to /srv/home/firdaus/.conda/envs/neon/lib/python2.7/site-packages
Adding pycuda 2015.1.3 to easy-install.pth file

Installed /srv/home/firdaus/.conda/envs/neon/lib/python2.7/site-packages/pycuda-2015.1.3-py2.7-linux-x86_64.egg
Processing dependencies for pycuda==2015.1.3
Searching for appdirs>=1.4.0
Reading https://pypi.python.org/simple/appdirs/
Download error on https://pypi.python.org/simple/appdirs/: [Errno 111] Connection refused -- Some packages may not be found!
Couldn't find index page for 'appdirs' (maybe misspelled?)
Scanning index of all packages (this may take a while)
Reading https://pypi.python.org/simple/
Download error on https://pypi.python.org/simple/: [Errno 111] Connection refused -- Some packages may not be found!
No local packages or download links found for appdirs>=1.4.0
error: Could not find suitable distribution for Requirement.parse('appdirs>=1.4.0')

```

**Comment by nervanasys** [comment]
We are looking into what's up with the Anaconda installation. In the mean time can you use the regular apt-get python ?

**Comment by magnord** [comment]
I'll do that. Thanks for looking into thus. 

I think another problem with the Anaconda install instructions (in addition to the dependency issues) is that skipping the "make" step makes no sense. A lot of stuff will never get built if you use those instructions.

**Comment by hanlin-nervana** [comment]
I've been able to configure neon for anaconda on a Ubuntu 14.04 Kepler instance on AWS with the following:

Create cuda environment variables (substitute your own proper cuda path)
```
export PATH="/usr/local/cuda/bin:$PATH"
export LD_LIBRARY_PATH="/usr/local/cuda/lib64:$LD_LIBRARY_PATH"
```

Then, create a conda environment
```
conda create --name neon pip
```

Activate the environment, then run a system-wide install
```
source activate neon
cd ~/path/to/neon
make sysinstall
```

Because you are inside a conda environment, sysinstall will install all the files in `/path-to-anaconda/envs/neon/bin/python`. I was able to run GPU examples without any issues. Let me know if that works!
